File: TSOS/JS/lab_4/src/lab/Correlation.ts
import * as math from 'mathjs';
import { Butterfly } from './Butterfly';

export class Сorrelation {
  public static CrossCorrelation(
    signal1: number[],
    signal2: number[]
  ): math.Complex[] {
    if (signal1.length != signal2.length) {
      alert('something went');
      throw new Error('The lengths of arrays must be the same');
    }

    let L = signal1.length + signal2.length;

    const correlation: math.Complex[] = new Array(L);

    let index1 = 0;
    let index2 = signal2.length - 1;

    let start;
    let end;
    let index;

    for (let i = 0; i < L; i++) {
      let sum: math.Complex = math.complex(0, 0);

      if (i < signal1.length - 1) {
        start = index2;
        end = signal2.length;
        index = 0;

        for (let j = start; j < end; j++) {
          const mutlplier = math.multiply(signal1[index++], signal2[j]);

          sum = math.add(sum, mutlplier) as math.Complex;
        }

        index1++;
        index2--;
      } else {
        start = signal1.length - index1 - 1;
        end = signal1.length;
        index = 0;

        for (let j = start; j < end; j++) {
          sum = math.add(
            sum,
            math.multiply(signal1[j], signal2[index++])
          ) as math.Complex;
        }

        index1--;
        index2++;
      }

      correlation[i] = sum;
    }

    return Сorrelation.Normalize(correlation);
  }

  public static FastCrossCorrelation(
    signal1: number[],
    signal2: number[]
  ): math.Complex[] {
    const L = signal1.length + signal2.length;

    const complexSignal1: math.Complex[] = new Array(L)
      .fill(undefined)
      .map(() => math.complex(0, 0));
    const complexSignal2: math.Complex[] = new Array(L)
      .fill(undefined)
      .map(() => math.complex(0, 0));

    for (let i = 0; i < signal1.length; i++) {
      complexSignal1[i] = math.complex(signal1[i], 0);
    }

    for (let i = 0; i < signal2.length; i++) {
      complexSignal2[i] = math.complex(signal2[i], 0);
    }

    const bpf1: math.Complex[] = Butterfly.DecimationInTime(
      complexSignal1,
      true
    );
    const bpf2: math.Complex[] = Butterfly.DecimationInTime(
      complexSignal2,
      true
    );

    const multiplicated: math.Complex[] = new Array(L);
    for (let i = 0; i < L; i++) {
      bpf1[i] = math.divide(bpf1[i], bpf1.length) as math.Complex;
      bpf2[i] = math.divide(bpf2[i], bpf2.length) as math.Complex;
      // @ts-ignore
      multiplicated[i] = math.multiply(bpf1[i], bpf2[i].conjugate()) as Complex;
    }

    const correlation: math.Complex[] = Butterfly.DecimationInTime(
      multiplicated,
      false
    );

    return Сorrelation.Normalize(correlation);
  }

  public static AutoCorrelation(
    signal: number[],
    shift: number
  ): math.Complex[] {
    return Сorrelation.CrossCorrelation(
      signal,
      Сorrelation.GetShiftedSignal(signal, shift)
    );
  }

  public static FastAutoCorrelation(
    signal: number[],
    shift: number
  ): math.Complex[] {
    return Сorrelation.FastCrossCorrelation(
      signal,
      Сorrelation.GetShiftedSignal(signal, shift)
    );
  }

  private static Normalize(values: math.Complex[]): math.Complex[] {
    const result: math.Complex[] = new Array(values.length);

    const max = Math.max(...values.map(e => e.re)); //values.Max(c => Math.abs(c.Real));

    for (let i = 0; i < result.length; i++) {
      result[i] = math.divide(values[i], max) as math.Complex;
    }

    return result;
  }

  private static GetShiftedSignal(signal: number[], shift: number): number[] {
    const shiftedSignal: number[] = new Array<number>(signal.length);

    for (let i = 0; i < shiftedSignal.length; i++) {
      let index = i - shift;
      if (index < 0) {
        index = signal.length + index;
      }
      shiftedSignal[i] = signal[index];
    }
    return shiftedSignal;
  }
}

File: TSOS/C#/lab_4/DSP.Lab4.Api/Correlation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;

namespace DSP.Lab4.Api
{
    public static class Сorrelation
    {
        public static Complex[] CrossCorrelation(double[] signal1, double[] signal2)
        {
            if (signal1.Length != signal2.Length)
            {
                throw new ArgumentException("The lengths of arrays must be the same");
            }

            int L = signal1.Length + signal2.Length;

            Complex[] correlation = new Complex[L];

            int index1 = 0;
            int index2 = signal2.Length - 1;

            int start;
            int end;
            int index;

            for (int i = 0; i < L; i++)
            {
                Complex sum = 0;

                if (i < signal1.Length - 1)
                {
                    start = index2;
                    end = signal2.Length;
                    index = 0;

                    for (int j = start; j < end; j++)
                    {
                        sum += signal1[index++] * signal2[j];
                    }

                    index1++;
                    index2--;
                }
                else
                {
                    start = signal1.Length - index1 - 1;
                    end = signal1.Length;
                    index = 0;

                    for (int j = start; j < end; j++)
                    {
                        sum += signal1[j] * signal2[index++];
                    }

                    index1--;
                    index2++;
                }

                correlation[i] = sum;
            }

            return Normalize(correlation);
        }

        public static Complex[] FastCrossCorrelation(double[] signal1, double[] signal2)
        {
            int L =  signal1.Length + signal2.Length;

            Complex[] complexSignal1 = new Complex[L];
            Complex[] complexSignal2 = new Complex[L];

            for (int i = 0; i < signal1.Length; i++)
            {
                complexSignal1[i] = signal1[i];
            }

            for (int i = 0; i < signal2.Length; i++)
            {
                complexSignal2[i] = signal2[i];
            }

            Complex[] bpf1 = Butterfly.DecimationInTime(complexSignal1, true);
            Complex[] bpf2 = Butterfly.DecimationInTime(complexSignal2, true);

            Complex[] multiplicated = new Complex[L];
            for (int i = 0; i < L; i++)
            {
                bpf1[i] /= bpf1.Length;
                bpf2[i] /= bpf2.Length;
                multiplicated[i] = bpf1[i] * Complex.Conjugate(bpf2[i]);
            }

            Complex[] correlation = Butterfly.DecimationInTime(multiplicated, false);

            //for (int i = 0; i < correlation.Length / 2; i++)
            //{
            //    Complex temp = correlation[i];
            //    correlation[i] = correlation[correlation.Length / 2 + i];
            //    correlation[correlation.Length / 2 + i] = temp;
            //}

            return Normalize(correlation);
        }

        public static Complex[] AutoCorrelation(double[] signal, int shift)
        {
            return CrossCorrelation(signal, GetShiftedSignal(signal, shift));
        }

        public static Complex[] FastAutoCorrelation(double[] signal, int shift)
        {
            return FastCrossCorrelation(signal, GetShiftedSignal(signal, shift));
        }

        private static Complex[] Normalize(Complex[] values)
        {
            Complex[] result = new Complex[values.Length];

            double max = values.Max(c => Math.Abs(c.Real));

            for (int i = 0; i < result.Length; i++)
            {
                result[i] = values[i] / max;
            }

            return result;
        }

        private static double[] GetShiftedSignal(double[] signal, int shift)
        {
            double[] shiftedSignal = new double[signal.Length];

            for (int i = 0; i < shiftedSignal.Length; i++)
            {
                int index = i - shift;
                if (index < 0)
                {
                    index = signal.Length + index;
                }
                shiftedSignal[i] = signal[index];
            }
            return shiftedSignal;
        }
    }
}
